File: pulsar-ql-server/src/main/java/fun/platonic/pulsar/ql/DbSession.java
package fun.platonic.pulsar.ql;

import org.h2.command.CommandInterface;

/**
 * DbSession is a wrapper for underlying database session, and is the bridge between database session and pulsar query
 * session
 * */
public class DbSession {
    private int sqlSequence;
    private int id;
    private String name;
    private Object implementation;

    public DbSession(Object implementation) {
        this.implementation = implementation;
        if (org.h2.engine.Session.class.isAssignableFrom(implementation.getClass())) {
            org.h2.engine.Session session = (org.h2.engine.Session)implementation;
            id = session.getId();
            name = session.toString();
        }
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getSqlSequence() {
        if (org.h2.engine.Session.class.isAssignableFrom(implementation.getClass())) {
            org.h2.engine.Session session = (org.h2.engine.Session)implementation;
            sqlSequence = session.getCommandSequence();
        }

        return sqlSequence;
    }

    public void setSqlSequence(int sqlSequence) {
        this.sqlSequence = sqlSequence;
    }

    public int executeUpdate(String sql) {
        if (org.h2.engine.Session.class.isAssignableFrom(implementation.getClass())) {
            org.h2.engine.Session session = (org.h2.engine.Session)implementation;
            CommandInterface command = session.prepareCommand(sql);
            return command.executeUpdate();
        }

        return 0;
    }

    @Override
    public int hashCode() {
        return id;
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof DbSession && ((DbSession) obj).id == id;
    }
}

File: pulsar-ql-server/src/main/kotlin/ai/platon/pulsar/ql/DbSession.kt
package ai.platon.pulsar.ql

/**
 * DbSession is a wrapper for underlying database session, it is a bridge between database session and pulsar query
 * session
 */
class DbSession(val id: Int, val implementation: Any) {
    var sqlSequence: Int = 0
        get() {
            if (org.h2.engine.Session::class.java.isAssignableFrom(implementation.javaClass)) {
                val session = implementation as org.h2.engine.Session
                this.sqlSequence = session.commandSequence
            }

            return field
        }
    var name: String = ""
        private set

    init {
        if (org.h2.engine.Session::class.java.isAssignableFrom(implementation.javaClass)) {
            val session = implementation as org.h2.engine.Session
            name = session.toString()
        }
    }

    fun executeUpdate(sql: String): Int {
        if (org.h2.engine.Session::class.java.isAssignableFrom(implementation.javaClass)) {
            val session = implementation as org.h2.engine.Session
            val command = session.prepareCommand(sql)
            return command.executeUpdate()
        }

        return 0
    }

    override fun hashCode(): Int {
        return id
    }

    override fun toString(): String {
        return name
    }

    override fun equals(obj: Any?): Boolean {
        return obj is ai.platon.pulsar.ql.DbSession && obj.id == id
    }
}
